keps/0010-dynamic-plugin-configuration.md

```diff
@@ -0,0 +1,221 @@
+difficult to manage. While it is possible to load configuration from a secret using
+the `configFrom` parameter, the entire configuration must be contained with in the
+secret which is effectively the same problem.
+
+It would be advantageous, and more in line with common Kubernetes behavior to be able
+to derive individual values of plugin configuration from data already stored on the same
+Kubernetes cluster. This proposal aims to define a change to the `KongPlugin` and `KongClusterPlugin`
+Custom Resource Definitions to facilitate this behavior.
+
+## Motivation
+
+Allow `KongPlugin` and `KongClusterPlugin` resources to read configuration values from Kubernetes,
+in addition to literal values.
+
+Continue to allow `KongPlugin` and `KongClusterPlugin` resource to accept literal input values.
+
+Reduce the amount of duplication of configuration value in the simple case. That is,
+when using standard kubernetes objects and with out assumed templating systems.
+
+### Goals
+
+* Provide ability to read individual configuration values from `ConfigMap` resources
+* Provide ability to read individual configuration values from `Secret` resources
+* Provide ability to read individual configuration values from arbitrary or generic resource types (other custom resource types)
+
+### Non-Goals
+
+* Introduction of a new Custom Resource or different Plugin Resource
+
+## Proposal
+
+In an effort to stay in line with current Kubernetes resource patterns it would be desirable
+to mimic the way pods define environment variables using an array of objects where
+the objects contain a `name` and one of `value` or `valueFrom`. `valueFrom`
+would further supports `configMapKeyRef`, `secretKeyRef` or `genericKeyRef`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a operator I want certain plugin configuration values to be dynamic in addition
+to statically definable
+
+#### Story 2
+
+As a operator I want to be able to derive a literal value using value interpolation referencing
+other values
+
+#### Story 3
+
+As an operator I want values stored in `ConfigMap` or `Secret` resources
+to be the source of truth for various resources deployed to my Kubernetes cluster
+
+#### Story 4
+
+As an operator I expect changes in `ConfigMap` or `Secret` data values to be reflected
+in the resource and deployments that depend on them with out having to restart critical parts of the system
+
+#### Story 5
+
+As an operator I expect optional dynamic values to support default values to be utilized
+when the key or resource is missing from the cluster
+
+### Notes/Constraints/Caveats (Optional)
+
+Using the data defined from the existing `config` values first and utilizing the
+`configFrom` values may provide a simple transition forward. This would allow existing
+configuration to continue to work as it currently does and values defined by `configFrom`
+override. Combined with a feature gate, it may provide means to move between styles
```

Comment:
An alternative to this we should consider is making a `v2` of the `KongPlugin` API, with a conversion webhook that can convert between them.

Reply:
That was my original thought / suggestion on #2763. @rainest indicated you all try to maintain backward compatibility, and that option would make that difficult. 

If it isn't the case - I'm very in favor of this option

Reply:
Sounds good, can we capture that perhaps in the notes or a TODO list (or have we already and I just have missed it)? I would consider this comment resolved as long as we have some kind of follow up on that for posterity.

Reply:
added as an alternative. And as additions to `notes`